models/GroupsService.php

```diff
@@ -74,48 +95,47 @@ public function deleteGroup(\core_kernel_classes_Resource $group)
      *
      * @access public
      * @author Joel Bout, <joel.bout@tudor.lu>
-     * @param  Class clazz
+     * @param  core_kernel_classes_Class $clazz
      * @return boolean
      */
     public function isGroupClass(core_kernel_classes_Class $clazz)
     {
-        return $clazz->equals($this->getRootClass()) || $clazz->isSubClassOf($this->getRootClass());
+        return $clazz->equals($this->getRootClass())
+            || $clazz->isSubClassOf($this->getRootClass());
     }
 
     /**
-     * get the groups of a user
+     * Get the groups of a user
      *
      * @access public
      * @author Joel Bout, <joel.bout@tudor.lu>
-     * @param  string userUri
+     * @param  User $user
      * @return array resources of group
      */
     public function getGroups(User $user)
     {
-        $groups = $user->getPropertyValues(self::PROPERTY_MEMBERS_URI);
-        array_walk($groups, function (&$group) {
-            $group = new core_kernel_classes_Resource($group);
-        });
-        return $groups;
+        return array_map(
+            function (string $group): core_kernel_classes_Resource {
+                return $this->getModel()->getResource($group);
+            },
+            $user->getPropertyValues(self::PROPERTY_MEMBERS_URI)
+        );
     }
-    
+
     /**
-     * gets the users of a group
+     * Gets the users of a group
      *
      * @param string $groupUri
-     * @return array resources of users
+     * @return core_kernel_classes_Resource[] resources of users
      */
-    public function getUsers($groupUri)
+    public function getUsers(string $groupUri): array
```

Comment:
Are we sure this one isn't used elsewhere?

Reply:
This is called from the new method `cloneInstance()` in this same class and from taoDeliveryRdf in a foreach loop (`GroupsService::singleton()->getUsers($group)`). The parameter name already suggests that the method expects a URI (a string) instead of something else, and the former PHPDoc states the parameter type is a string.

I've not found any method that overloads this one, so I think adding the type hint for this parameter is safe.

Reply:
We just need to check also if some calls with null could happen.